habitat-lab/habitat/datasets/rearrange/navmesh_utils.py

```diff
@@ -171,6 +177,203 @@ def unoccluded_navmesh_snap(
+                and agent_embodiment is None
+            ):
+                # No embodiment for collision detection, so return closest unoccluded point
+                return batch_sample[0], facing_target_angle, True
+
+            # get orientation noise offset
+            orientation_noise_samples = []
+            if orientation_noise > 0 and max_orientation_samples > 0:
+                orientation_noise_samples = [
+                    np.random.normal(0.0, orientation_noise)
+                    for _ in range(max_orientation_samples)
+                ]
+            # last one is always no-noise to check forward-facing
+            orientation_noise_samples.append(0)
+
+            for orientation_noise_sample in orientation_noise_samples:
+                desired_angle = facing_target_angle + orientation_noise_sample
+                if embodiment_heuristic_offsets is not None:
+                    # local 2d point rotation
+                    rotation_2d = mn.Matrix3.rotation(-mn.Rad(desired_angle))
+                    transformed_offsets_2d = [
+                        rotation_2d.transform_vector(xz)
+                        for xz in embodiment_heuristic_offsets
+                    ]
+
+                    # translation to global 3D points at navmesh height
+                    offsets_3d = [
+                        np.array(
+                            [
+                                transformed_offset_2d[0],
+                                0,
+                                transformed_offset_2d[1],
+                            ]
+                        )
+                        + batch_sample[0]
+                        for transformed_offset_2d in transformed_offsets_2d
+                    ]
+
+                    if data_out is not None:
+                        data_out["offsets_3d"] = offsets_3d
+
+                    # check for offset navigability
+                    is_collision = False
+                    for offset_point in offsets_3d:
+                        if not (
+                            sim.pathfinder.is_navigable(offset_point)
+                            and (
+                                island_id == -1
+                                or sim.pathfinder.get_island(offset_point)
+                                == island_id
+                            )
+                        ):
+                            is_collision = True
+                            break
+
+                    # if this sample is invalid, try the next
+                    if is_collision:
+                        continue
+
+                if agent_embodiment is not None:
+                    # contact testing with collision shapes
+                    start_position = agent_embodiment.base_pos
+                    start_rotation = agent_embodiment.base_rot
+
+                    agent_embodiment.base_pos = batch_sample[0]
+                    agent_embodiment.base_rot = desired_angle
```

Comment:
I am not sure I understand orientation noise in these functions. Should it apply to desired_angle here/agent orientation basically?

Reply:
Yeah, basically this is an option to sample random noisy orientations from a gaussian centered on the forward facing direction (robot facing toward the target).
The final, valid orientation is part of the return Tuple.

We have to set the agent orientation during the sampling process such that the collision checking can validate it.

This is helpful because it may be the case that some parts of the scene are tight enough that the robot will only fit into the space within a limited range of orientations (e.g. walking between sofa and coffee table). Sampling accurately in these kinds of poses requires many iterations and a high degree of noise.